Post/troubleshooting nuget package (#269)
adessoSE/devblog: _posts/2022-12-16-Troubleshooting-open-source-NuGet-package.md

   1022222:	 search cache=/etc/ld.so.cache
   1022222:	 search path=/usr/lib/tls:/usr/lib		(system search path)
   1022222:	  trying file=/usr/lib/tls/libgit2-106a5f2.so
   1022222:	  trying file=/usr/lib/libgit2-106a5f2.so
   1022222:
   1022222:	[...]
   1022222:	find library=git2-106a5f2 [0]; searching
   1022222:	 search cache=/etc/ld.so.cache
   1022222:	 search path=/usr/lib/tls:/usr/lib		(system search path)
   1022222:	  trying file=/usr/lib/tls/git2-106a5f2
   1022222:	  trying file=/usr/lib/git2-106a5f2
   1022222:
   1022222:	find library=libgit2-106a5f2 [0]; searching
   1022222:	 search cache=/etc/ld.so.cache
   1022222:	 search path=/usr/lib/tls:/usr/lib		(system search path)
   1022222:	  trying file=/usr/lib/tls/libgit2-106a5f2
   1022222:	  trying file=/usr/lib/libgit2-106a5f2
   1022222:
Unhandled exception.
System.TypeInitializationException: The type initializer for 'LibGit2Sharp.Core.NativeMethods' threw an exception.
 ---> System.DllNotFoundException: Unable to load shared library 'git2-106a5f2' or one of its dependencies.
      [...]
make: *** [Makefile:18: run] Error 134
```
_Fig4: LD_DEBUG run to list the libraries_

> When setting the `LD_DEBUG` parameter with `all` or `files` we get similar results:

Author: Danke für die ganzen Verbesserungen.
Sollte das Datum für den Post noch aktualisiert werden oder passt der Wert (`2022-12-16 12:00`)?

> When setting the `LD_DEBUG` parameter with `all` or `files` we get similar results:

Reviewer: Das Datum wird von uns direkt vor dem Merge gesetzt, damit der Post dann als aktuellster im Blog angezeigt wird.

   1022654:	file=/home/vince/Projects/poc_libgit2sharp/poc_libgit2sharp/bin/Debug/net6.0/runtimes/linux-x64/native/libgit2-106a5f2.so [0];  generating link map
   1022654:	  dynamic: 0x00007fda6ad24d90  base: 0x00007fda6aa00000   size: 0x000000000032a5d0
   1022654:	    entry: 0x00007fda6aa166f0  phdr: 0x00007fda6aa00040  phnum:                  7
```
_Fig5: Output from `LD_DEBUG=files`_

The library files `./bin/Debug/net6.0/runtimes/linux-x64/native/libgit2-106a5f2.so` does exist and can be imported correctly.
On the other hand the referenced file `/usr/share/dotnet/shared/Microsoft.NETCore.App/6.0.2/git2-106a5f2.so` does not exist at this global location.

If there would be no specific version applied to the shared object file I would say the file should come from the system.
Let's check the package manager and install the latest version of the underlying library `libgit2` on the system.

``` bash
$ sudo pacman -S libgit2
```
_Fig6: Installing the package globally [packages libgit2](https://archlinux.org/packages/extra/x86_64/libgit2/)_

Even after the installation/update the error persists.

> So as expected there is no correlation with the global installation of the library and it should come all bundled with the NuGet.

Reviewer: ```suggestion
So, as expected, there is no correlation with the global installation of the library and it should come all bundled with the NuGet package.
```

Going further and exploring the issues on GitHub and discovering a potential fix will be the next step.

# Exploring GitHub issues for potential fixes

Before exploring the installation scripts or the code of the library let us explore the issues on GitHub.

> There are some issues hinting that the correct dependency ob the package should be used which is `LibGit2Sharp.NativeBinaries` in version `2.0.306` which was already the case when inspecting the file `./obj/project.assets.json`.

Reviewer: ```suggestion
There are some issues hinting that the correct dependency of the package should be used, which is `LibGit2Sharp.NativeBinaries` in version `2.0.306`, which was already the case when inspecting the file `./obj/project.assets.json`.
```

> On the other hand there are multiple issues indicating that there are issues with .Net 5 which should be fixed with the preview version of `0.27.0` of the `LibGit2Sharp` NuGet.

Reviewer: ```suggestion
On the other hand, there are multiple issues indicating that there are issues with .Net 5, which should be fixed with the preview version of `0.27.0` of the `LibGit2Sharp` package.
```

## Installing the preview version of the library

Now lets just update the package reference to check if the preview version will fix it for our .Net 6 application.

``` xml
  <ItemGroup>
    <PackageReference Include="LibGit2Sharp" Version="0.27.0-preview-0182" />
  </ItemGroup>
```
_Fig7: [poc_libgit2sharp.csproj](https://git.sr.ht/~vince/poc_libgit2sharp/tree/b10eb21f98dcffe40dcfcd792933a795bbcff3dd/item/poc_libgit2sharp/poc_libgit2sharp.csproj)_

``` bash
~/Projects/poc_libgit2sharp main $ make run
[...]
dotnet run --project ./poc_libgit2sharp/poc_libgit2sharp.csproj
Current directory is '/home/vince/Projects/poc_libgit2sharp'
Checking whether the directory contains a valid git repository.
Should return true as the program is run from a repository.
This message indicates that no exception was thrown and the library works as expected.
The directory '/home/vince/Projects/poc_libgit2sharp' is a git repository 'True'
```
_Fig8: Execute with preview version_

Perfect, now no exception gets thrown from the library and the expected result holds true.

> The working directory is indeed a git [repository](https://git.sr.ht/~vince/poc_libgit2sharp).

Reviewer: ```suggestion
The working directory is indeed a [git repository](https://git.sr.ht/~vince/poc_libgit2sharp).
```

> Now having a reference to a preview version is always unpleasant and should be avoided.

Reviewer: ```suggestion
Now, having a reference to a preview version is always unpleasant and should be avoided.
```

> As we know with git itself we have the possibility of including the sources directly and building it with our project.

Reviewer: ```suggestion
As we know with git itself, we have the possibility of including the sources directly and building them within our project.
```

> To draw a conclusion for the project lets explore both options in this MVP.

Reviewer: ```suggestion
To draw a conclusion for the project, lets explore both options in this MVP.
```

# Building from source

> Now for the last step lets try to build the project ourselves and include the result in our project.

Reviewer: ```suggestion
Now for the last step, lets try to build the project ourselves and include the result in our project.
```

> With this option we can directly link the `master` branch of the repository and update it accordingly to receive the latest updates instead of waiting for an update on the NuGet package.

Reviewer: ```suggestion
With this option, we can directly link the `master` branch of the repository and update it accordingly to receive the latest updates instead of waiting for an update on the NuGet package.
```

``` bash
$ git submodule add https://github.com/libgit2/libgit2sharp.git libgit2sharp
[...]

$ ls -l
-rw-r--r--  1 vince vince  349 Apr 21 20:23 .gitignore
-rw-r--r--  1 vince vince   99 Apr 21 22:16 .gitmodules
[...]
drwxr-xr-x 9 vince vince  4096 Apr 21 22:24 libgit2sharp
-rw-r--r-- 1 vince vince   432 Apr 21 20:54 Makefile
drwxr-xr-x 4 vince vince  4096 Apr 21 20:21 poc_libgit2sharp

$ cat .gitmodules
[submodule "libgit2sharp"]
	path = libgit2sharp
	url = https://github.com/libgit2/libgit2sharp.git
```
_Fig9: Installing the submodule_

> And finally update the project reference.

Reviewer: ```suggestion
And finally, we can update the project reference.
```

``` bash
$ git diff d85adbb~..d85adbb
diff --git a/poc_libgit2sharp/poc_libgit2sharp.csproj b/poc_libgit2sharp/poc_libgit2sharp.csproj
index 4329b2b..f3f8c62 100644
--- a/poc_libgit2sharp/poc_libgit2sharp.csproj
+++ b/poc_libgit2sharp/poc_libgit2sharp.csproj
@@ -8,7 +8,7 @@
   </PropertyGroup>

   <ItemGroup>
-    <PackageReference Include="LibGit2Sharp" Version="0.27.0-preview-0182" />
+    <ProjectReference Include="..\libgit2sharp\LibGit2Sharp\LibGit2Sharp.csproj" />
   </ItemGroup>

 </Project
```
_Fig10: Update the reference_

> The submodule works as expected just as the preview version of the NuGet.

Reviewer: ```suggestion
The submodule works as expected just as the preview version of the NuGet package.
```

> With the submodule we have similar downsides to using a package manager.
> As the reference has to be updated manually when a new version is published.

Reviewer: ```suggestion
With the submodule, we have similar downsides to using a package manager, as the reference has to be updated manually when a new version is published.
```

> When using the submodule we can directly checkout the `master` branch and check the latest development.

Reviewer: ```suggestion
When using the submodule, we can directly checkout the `master` branch and check out the latest development version.
```

It is worth noting that when a stable release of `0.27.0` is released of the package this choice has to be reevaluated.

# Conclusion

> To see all the changes made since the last published version we can checkout the diff on GitHub directly [v0.26..master](https://github.com/libgit2/libgit2sharp/compare/v0.26..master).

Reviewer: ```suggestion
To see all the changes made since the last published version, we can check out the diff on GitHub directly [v0.26..master](https://github.com/libgit2/libgit2sharp/compare/v0.26..master).
```

> The help on submodules is extensive and can be viewed under `man 1 git-submodule` and `man 7 gitsubmodules`.

Reviewer: Ist es beabsichtigt, dass da einmal git-submodules und einmal gitsubmodules steht?

Author: Ja das ist beabsichtigt.
Es gibt in Sektion 7 einen Eintrag für `gitsubmodules` und in Sektion 1 für `git-submodule`.

> For my project I decided to go with the submodule since it gives me a few advantages:

Reviewer: ```suggestion
For my project I decided to go with the submodule workaround since it gives me a few advantages:
```

> * Building regularly let me observe the submodule for updates

Reviewer: ```suggestion
* Building regularly lets me observe the submodule for updates
```

* I can keep up with changes to the library

> * Give early feedback as soon as I will notice it

Reviewer: Was möchtest du hier sagen, der Satz ergibt so keinen Sinn.
